### ISSUE TITLE
user's permissions on Issues/PRs

### ISSUE THREAD

**Issue by marusinm** [packit-service/ogr#100]
ogr.abstract.Issue and ogr.abstract.PullRequest doesn't contain a field if a user has permission to close Issue or merge PR. 

It should be detected by the user's token. In the case of Github, roles of `owner`, `member`, `collaborator` should have permissions. 

It could be related to #88 if we save a list of owners, members, and collaborators into GitProject instance or create get methods for them.

**Comment by lachmanfrantisek** [comment]
Isn't this already fixed for GitHub and Pagure?

**Comment by lachmanfrantisek** [comment]
- We have it in Gitlab as well.
- We have `can_merge_pr`/`who_can_merge_pr`/`who_can_close_issue` methods in the project classes.

Can we close this issue?

Or, do we want/need related methods in Issue/PR classes as well? (`my_issue.who_can_close`/`issue_can_close("username")`)

**Comment by lachmanfrantisek** [comment]
I've updated the description. This issue is now about cleaning the methods and add the missing parts. Please, check if I don't forget anything.

**Comment by pranav33317** [comment]
Yes ma'am , I am working on it , will draft a PR ASAP . Had a question about the unimplemented functions , a lot of the functions are missing return statements and thus giving error . Should I just return default value or raise some error ?
Thanks

**Comment by lbarcziova** [comment]
@pranav33317 hello and thanks! 

> a lot of the functions are missing return statements and thus giving error .

Could you please provide some examples?

**Comment by pranav33317** [comment]
def get_issue_list(
        self,
        status: IssueStatus = IssueStatus.open,
        author: Optional[str] = None,
        assignee: Optional[str] = None,
        labels: Optional[list[str]] = None,
    ) -> list[Issue]:
        pass

    @indirect(GithubIssue.get)
    def get_issue(self, issue_id: int) -> Issue:
        pass

    @indirect(GithubIssue.create)
    def create_issue(
        self,
        title: str,
        body: str,
        private: Optional[bool] = None,
        labels: Optional[list[str]] = None,
        assignees: Optional[list[str]] = None,
    ) -> Issue:
        pass , @indirect(GithubPullRequest.get_list)
    def get_pr_list(self, status: PRStatus = PRStatus.open) -> list[PullRequest]:
        pass

    @indirect(GithubPullRequest.get)
    def get_pr(self, pr_id: int) -> PullRequest:
        pass , def create_pr(
        self,
        title: str,
        body: str,
        target_branch: str,
        source_branch: str,
        fork_username: Optional[str] = None,
    ) -> PullRequest:
        pass, def set_commit_status(
        self,
        commit: str,
        state: Union[CommitStatus, str],
        target_url: str,
        description: str,
        context: str,
        trim: bool = False,
    ):
        pass

    @indirect(GithubCommitFlag.get)
    def get_commit_statuses(self, commit: str) -> list[CommitFlag]:
        pass

    @indirect(GithubCheckRun.get)
    def get_check_run(
        self,
        check_run_id: Optional[int] = None,
        commit_sha: Optional[str] = None,
    ) -> Optional["GithubCheckRun"]:
        pass

    @indirect(GithubCheckRun.create)
    def create_check_run(
        self,
        name: str,
        commit_sha: str,
        url: Optional[str] = None,
        external_id: Optional[str] = None,
        status: GithubCheckRunStatus = GithubCheckRunStatus.queued,
        started_at: Optional[datetime.datetime] = None,
        conclusion: Optional[GithubCheckRunResult] = None,
        completed_at: Optional[datetime.datetime] = None,
        output: Optional[GithubCheckRunOutput] = None,
        actions: Optional[list[dict[str, str]]] = None,
    ) -> "GithubCheckRun":
        pass

    @indirect(GithubCheckRun.get_list)
    def get_check_runs(
        self,
        commit_sha: str,
        name: Optional[str] = None,
        status: Optional[GithubCheckRunStatus] = None,
    ) -> list["GithubCheckRun"]:
        pass
, these are some of the examples of unimplemented functions missing return statements , they are mostly in the following files : ogr/services/github/project.py and ogr/services/gitlab/project.py

**Comment by lbarcziova** [comment]
and where do you see these causing errors? From what you posted I would say the methods are implemented, just not directly, but using the `@indirect` decorator, with methods on the particular objects.

**Comment by pranav33317** [comment]
All of them have empty bodies and require a return statement so they are returning error : 

<img width="582" alt="Image" src="https://github.com/user-attachments/assets/74a0d7ed-4b86-4691-ae83-283d4b9aa1f7" />

<img width="585" alt="Image" src="https://github.com/user-attachments/assets/ccf28b70-94d0-4bf8-a334-c8b920f1f38d" />

**Comment by pranav33317** [comment]
Hi , I'm sorry I commented on the wrong issue . This is related to https://github.com/packit/ogr/issues/696 , I will repost it there.
I'm working on this issue parallely as well.

**Comment by Gauravpadam** [comment]
The description is helpful. I'm looking into this @lachmanfrantisek

**Comment by Gauravpadam** [comment]
If I understand this correctly:

- `abstract.py` contains all the abstractions for the class methods. We are supposed to have all the mentioned methods as abstract methods first.
- Once that is done, We can actually start implementing these methods into their respective SubClasses in `github.py`, `gitlab.py` and `pagure.py`.
- We will write some unit tests for these changes.

Am I thinking in the right direction @lachmanfrantisek @lbarcziova?

If yes, I'll draft a PR for this ticket

**Comment by pranav33317** [comment]
Ma'am I was waiting for a response , could you please clear up my doubt , I was working on this

**Comment by lbarcziova** [comment]
@pranav33317 sorry, I got confused since you were mentioning #696 in the last comment. So what's the current status for this issue, are you blocked on anything at the moment? (I understood https://github.com/packit/ogr/issues/100#issuecomment-2742623129 is related to #696)

**Comment by pranav33317** [comment]
Yes ma'am still working on both . Could you please answer my query in #696 ?

**Comment by nforro** [comment]
Still relevant.
